editor/components/observe-typing/index.js

```diff
@@ -0,0 +1,166 @@
+/**
+ * External dependencies
+ */
+import { includes } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { Component, compose } from '@wordpress/element';
+import { withSelect, withDispatch } from '@wordpress/data';
+import { isTextField, keycodes } from '@wordpress/utils';
+
+const { UP, RIGHT, DOWN, LEFT, ENTER, BACKSPACE } = keycodes;
+
+/**
+ * Set of key codes upon which typing is to be initiated on a keydown event.
+ *
+ * @type {number[]}
+ */
+const KEY_DOWN_ELIGIBLE_KEY_CODES = [ UP, RIGHT, DOWN, LEFT, ENTER, BACKSPACE ];
+
+/**
+ * Returns true if a given keydown event can be inferred as intent to start
+ * typing, or false otherwise. A keydown is considered eligible if it is a
+ * text navigation without shift active.
+ *
+ * @param {KeyboardEvent} event Keydown event to test.
+ *
+ * @return {boolean} Whether event is eligible to start typing.
+ */
+function isKeyDownEligibleForStartTyping( event ) {
+	const { keyCode, shiftKey } = event;
+	return ! shiftKey && includes( KEY_DOWN_ELIGIBLE_KEY_CODES, keyCode );
+}
+
+class ObserveTyping extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.stopTypingOnSelectionUncollapse = this.stopTypingOnSelectionUncollapse.bind( this );
+		this.stopTypingOnMouseMove = this.stopTypingOnMouseMove.bind( this );
+		this.startTypingInTextField = this.startTypingInTextField.bind( this );
+
+		this.lastMouseMove = null;
+	}
+
+	componentDidMount() {
+		this.toggleEventBindings( this.props.isTyping );
+	}
+
+	componentDidUpdate( prevProps ) {
+		if ( this.props.isTyping !== prevProps.isTyping ) {
+			this.toggleEventBindings( this.props.isTyping );
+		}
+	}
+
+	componentWillUnmount() {
+		this.toggleEventBindings( false );
+	}
+
+	/**
+	 * Bind or unbind events to the document when typing has started or stopped
+	 * respectively, or when component has become unmounted.
+	 *
+	 * @param {boolean} isBound Whether event bindings should be applied.
+	 */
+	toggleEventBindings( isBound ) {
+		const bindFn = isBound ? 'addEventListener' : 'removeEventListener';
+		document[ bindFn ]( 'selectionchange', this.stopTypingOnSelectionUncollapse );
```

Comment:
I'm trying to recall if there was a specific reason to make this bound to the `document`, vs. just an `onSelectionChange` event on the `div` wrapper. It was inherited from how this was applied previously, originally from #4872, where the selection handling moved from `RichText` to `BlockListBlock`.

The document-based `mousemove` event makes sense because we want to unset typing even when the mouse is moved outside the descendents of this component. But `selectionchange`? I'm not sure. That seems inconsistent with the rest of the key handling that happens here.

Reply:
Ah, this would probably explain why:

https://github.com/facebook/react/issues/5785